tests/v2/test_1295-rdataframe-to-awkward.py

```diff
@@ -3,31 +3,83 @@
+
+import awkward._v2._connect.rdataframe._from_rdataframe  # noqa: E402
 
 
 ROOT = pytest.importorskip("ROOT")
 
-def test_from_rdf_to_awkward():
-    rdf = ROOT.RDataFrame(100)
-    array = rdf.AsAwkward()
-    assert array.tolist() == []
+compiler = ROOT.gInterpreter.Declare
 
-    rdf_x = rdf.Define("x", "gRandom->Rndm()")
-    array_x = rdf_x.AsAwkward()
-    assert (
-        str(array_x.form)
-        == """{
-    "class": "RecordArray",
-    "contents": {
-        "x": "float64"
-    }
-}"""
+
+def test_array_builder():
+    import ctypes
+
+    builder = ak.ArrayBuilder()
+
+    ROOT.gInterpreter.Declare(
+        f"""
+    #include <functional>
+
+    typedef uint8_t (*FuncPtr)(void*);
+    typedef uint8_t (*FuncIntPtr)(void*, int64_t);
+
+    uint8_t
+    test_beginlist() {{
+        return std::invoke(reinterpret_cast<FuncPtr>(reinterpret_cast<long>({ctypes.cast(ak._libawkward.ArrayBuilder_beginlist, ctypes.c_voidp).value})), reinterpret_cast<void *>({builder._layout._ptr}));
+    }}
+
+    uint8_t
+    test_endlist() {{
+        return std::invoke(reinterpret_cast<FuncPtr>(reinterpret_cast<long>({ctypes.cast(ak._libawkward.ArrayBuilder_endlist, ctypes.c_void_p).value})), reinterpret_cast<void *>({builder._layout._ptr}));
+    }}
+
+    uint8_t
+    test_integer(int64_t x) {{
+        return std::invoke(reinterpret_cast<FuncIntPtr>(reinterpret_cast<long>({ctypes.cast(ak._libawkward.ArrayBuilder_integer, ctypes.c_void_p).value})), reinterpret_cast<void *>({builder._layout._ptr}), x);
+    }}
+    """
+    )
+    ROOT.test_beginlist()
+    ROOT.test_integer(1)
+    ROOT.test_integer(2)
+    ROOT.test_integer(3)
+    ROOT.test_endlist()
+
+    assert ak.to_list(builder.snapshot()) == [[1, 2, 3]]
+
+    ROOT.test_beginlist()
+    ROOT.test_integer(1)
+    ROOT.test_integer(2)
+    ROOT.test_integer(3)
+    ROOT.test_endlist()
+
+    assert ak.to_list(builder.snapshot()) == [[1, 2, 3], [1, 2, 3]]
+
+
+def test_array_builder_root():
+    builder = ak.ArrayBuilder()
+    func = ak._v2._connect.rdataframe._from_rdataframe.connect_ArrayBuilder(
+        compiler, builder
     )
 
-    array_none = rdf_x.AsAwkward(exclude=["x"])
-    assert array_none.to_list() == []
+    getattr(ROOT, func["beginlist"])()
+    getattr(ROOT, func["integer"])(1)
+    getattr(ROOT, func["integer"])(2)
+    getattr(ROOT, func["integer"])(3)
+    getattr(ROOT, func["endlist"])()
+    getattr(ROOT, func["real"])(3.3)
+
+    assert ak.to_list(builder.snapshot()) == [[1, 2, 3], 3.3]
 
-def test_vectors():
-    list = []
-    print(ROOT.is_iterable(list))
+
+@pytest.mark.skip(reason="Failed to instantiate 'Foreach(function)'")
+def test_as_ak_array():
+    builder = ak.ArrayBuilder()
+    func = ak._v2._connect.rdataframe._from_rdataframe.connect_ArrayBuilder(
+        compiler, builder
+    )
+
+    rdf = ROOT.RDataFrame(100)
+    rdf_x = rdf.Define("x", "gRandom->Rndm()")
+    rdf_x.Foreach(lambda x: getattr(ROOT, func["real"])(x))
```

Comment:
@eguiraud - does `Foreach` work in Python? I cannot find any examples. Thanks

Reply:
Hi Ianna, you can call `Foreach` from Python but like with `Define`, `Filter` etc. you cannot pass a Python lambda, it has to be a C++ callable (and differently from those interfaces, `Foreach` does not have an overload that takes a string).

For example this works:

```python
import ROOT

ROOT.gInterpreter.Declare("""
struct MyFunctor {
    int operator()(ULong64_t e) { return e % 2; }
};
""")

f = ROOT.MyFunctor()
ROOT.RDataFrame(10).Define("x", f, ["rdfentry_"]).Display("x").Print()
```

Reply:
Thanks, @eguiraud. Am I doing something wrong?
```python
        f = ROOT.MyFunctor()
>       ROOT.RDataFrame(10).Define("x", "gRandom->Rndm()").Foreach("x", f).Display("x").Print()
E       TypeError: Template method resolution failed:
E         Failed to instantiate "Foreach(std::string,MyFunctor&)"
E         Failed to instantiate "Foreach(std::string,MyFunctor*)"
E         Failed to instantiate "Foreach(std::string,MyFunctor)"

f          = <cppyy.gbl.MyFunctor object at 0x7fcaced31510>

```

Reply:
I think it's `Foreach(f, ["x"])`?

Reply:
hey, `Foreach` returns nothing (you can't keep the call chain going afterwards) and does not define a column. The signature is what @agoose77 suggests. In @ianna 's code above you would actually use a `Define`, probably:

```
ROOT.RDataFrame(10).Define("x", "gRandom->Rndm()").Define("y", f, ["x"]).Display("x").Print()
```

Reply:
Ok, great, thanks! I think, I do want `Foreach`, because it does conversion to an awkward array (using an `ArrayBuilder`). However, it might be useful to be able to feed it back to RDF as a newly defined column... though, for now, I think it should be a different step.

Reply:
I think we want `Foreach` because ArrayBuilder and LayoutBuilder construct arrays as a side-effect, and `Foreach` is for generating side-effects.

Perhaps we should define an `AsAwkward` method on the RDataFrame Python object that is implemented in terms of `Foreach`. This method could create the ArrayBuilder/LayoutBuilder, generate the C++ code that fills it, run that code with `Foreach`, and return the ArrayBuilder/LayoutBuilder's `snapshot`.

Or maybe one ArrayBuilder per slot in `ForeachSlot`, and then `ak.concatenate` their snapshots at the end. (Sorry for the differing terminology: "snapshot" in Awkward is turning an ArrayBuilder/LayoutBuilder into a finalized Array, nothing to do with disk I/O.)

Reply:
> it might be useful to be able to feed it back to RDF as a newly defined column

If anyone wants to send data from one RDF node to another, there's already better ways of doing that (e.g. `Define`) than creating an Awkward Array with ArrayBuilder/LayoutBuilder and then feeding the resulting ak.Array back in. I would view this `AsAwkward` purely as an RDF sink.